crates/forge_app/src/tool_registry.rs

```diff
@@ -149,6 +157,21 @@ impl<S> ToolRegistry<S> {
         }
         Ok(())
     }
+
+    /// Determines if a tool should have a timeout applied.
+    /// According to issue #1402, only search, shell, fetch, and mcp tools
+    /// should timeout. Interactive operations (like permission requests)
+    /// should not timeout.
+    fn should_apply_timeout(tool_name: &ToolName) -> bool {
+        let name = tool_name.as_str();
+        matches!(
+            name,
+            "forge_tool_fs_search" |     // search operations
+            "forge_tool_process_shell" | // shell operations  
+            "forge_tool_net_fetch" /* fetch operations
+                                    * MCP tools are handled separately and always have timeouts */
+        )
+    }
```

Comment:
This is a very untyped implementation. I think we should put a basic timeout right before calling the service that's associated for that tool call.